Initial toHaveFormValues implementation (closes #60)

gnapse opened the pull request:
**What**:

Allow tests to assert that a form (or part of it) contains certain elements with certain corresponding values.

**Why**:

See #60 

**How**:

By gathering all form values under the given container, and matching it against the expected values.

Additional efforts were made to ensure a consistent gathering of form values to match more what a user interprets out of them.

- Handles single vs multiple selection `<select>`
- A set of radio buttons with the same name behave as a single select
- A set of checkboxes with the same name behave as a multi select
- `<input type="number">` is matched as a numeric value, even though internally is a string
- Several elements with the same name are grouped as an array value

**Checklist**:
<!-- Have you done all of these things?  -->

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

* [x] Documentation
* [x] Tests
* [x] Updated Type Definitions
* [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->

gnapse commented:
This is WIP. I still have to make it so it works within a certain form, and not other types of elements. It also lacks documentation and type definitions. But it's ready to be reviewed, so let me know if there's any feedback about the implementation.

kentcdodds commented on src/to-have-form-values.js:
```diff
@@ -0,0 +1,130 @@
+  return self.indexOf(value) === index
+}
+
+function isEqual(a, b) {
```
Maybe we should just use `lodash.isequal`

gnapse replied:
Yes, I initially resisted to add lodash as a dependency, but as I needed more and more, I considered it.

kentcdodds replied:
I often resist installing lodash, but then I remind myself that the chances someone doesn't already have lodash installed in their `node_modules` are very very slim 😅

gnapse replied:
However, there's still something else being done here besides a mere `isEqual`. Because we need to compare arrays as sets, not as arrays. Order is not important. But anyway, lodash will surely help with that anyway. This was indeed a sort of hack to get this going.

kentcdodds commented on src/to-have-form-values.js:
```diff
@@ -0,0 +1,130 @@
+}
+
+function getAllFormValues(container) {
+  const names = [...container.querySelectorAll('[name]')]
```
Rather than this query selector, why don't we rely more on the DOM's built-in mechanism for form elements? This would have the added benefit of not working if the form controls are not wired up to the form properly.

So rather than `container.querySelectorAll('[name]')` where `container` is actually `form`, we could just do: `form.elements`?

gnapse replied:
Definitely. As stated in the description, I've yet to check that the container is a form element or a field set.

gnapse commented:
Fixed the two pending observations (use of lodash and form.elements instead of rolling our own solution for those needs).

Still pending to add documentation and type definitions.

Also, there's an issue in #64 that is preventing me from actually checking if the element is an `instance HTMLFormElement` or of `HTMLFieldSetElement`. I added a less reliable test for this for the time being, until that issue is resolved.

gnapse commented:
Added documentation and type definition.

Not sure if the documentation is too large, but I felt I had to document all cases that this matcher covers. Regardless, feedback on the documentation would be great.

I also could not give a more accurate type definition. I would've liked to be more precise with something like this:

```typescript
toHaveFormValues({
  [name: string]: any
})
```

But for some reason I keep getting a TypeScript syntax error, when in other contexts I've been able to use this just fine. But I figure `toHaveFormValues(object)` is fine too.

Finally, I wonder if it makes sense to validate and error out if the given object is empty. It makes little sense to call `toHaveFormValues({})` but right now it will always pass. Guess that's no big deal either.

jgoz commented on extend-expect.d.ts:
```diff
@@ -12,11 +12,12 @@ declare namespace jest {
     toHaveAttribute(attr: string, value?: string): R
     toHaveClass(...classNames: string[]): R
+    toHaveFocus(): R
+    toHaveFormValues(object)
```
@gnapse I think the reason you were seeing syntax errors is because the parameter needs to be named. As it stands, this is also incorrect syntax.

```ts
toHaveFormValues(values: { [name: string]: any })
```

gnapse replied:
Thanks, I knew I must have been doing something wrong. I guess I need to take a break ;)

jgoz replied:
Also, it should specify a return value of `R`

gnapse commented:
A use case I wanted to bring to others' attention: what happens if there are multiple elements with the same name (other than radio buttons or checkboxes)? How should we treat those?

Right now I considered grouping them all as a single form control that returns a value that is an array of all the values of all the elements with that name. I only y do this if they're also all of the same `type`.

But I wonder if even if they're of the same type, I should do something different, such as failing with an error, or just defaulting to returning the value of only one of them, say the first or the last one. I think this is what submitting such a form would regularly do. But I also feel that we shouldn't be silent about it.

Thoughts?

kentcdodds commented:
What would happen in an old-school application that uses a form POST of the data? My guess is the later form control wins and it doesn't convert to an array. Could you validate that and let's just do what the browser would do.

gnapse commented:
In any case I think this matcher is a functionality that will probably get a lot of feedback in form of issues, bug reports, feature requests, after we release it, especially around how it treats edge or unconventional cases.

gnapse commented:
> What would happen in an old-school application that uses a form POST of the data? My guess is the later form control wins and it doesn't convert to an array. Could you validate that and let's just do what the browser would do.

I will do that. That's what I was inclined to do after thinking about this after I implemented the current approach.

I think I'll also print out a warning of some sort. Not making the matcher fail if there really isn't a mismatch. But also not let this unconventional condition pass through unnoticed. This is not a valid use case after all, and users may get failing tests inadvertently because of this.

kentcdodds commented:
If there's a log, there should be a way to disable the log 👍

gnapse commented:
Update: it seems to be valid to have multiple elements with the same name, not just radio buttons and checkboxes. Moreover, they seem to interpreted as an array, so I guess the current implementation is ok. Source: https://stackoverflow.com/questions/2906793/is-it-valid-to-have-two-input-elements-with-the-same-name

If agreed, I think this is ready to be reviewed and released if ok.

gnapse commented:
:tada: This PR is included in version 2.1.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/gnapse/jest-dom/releases/tag/v2.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
